Social | Open share post modal via URL query param

manzoorwanijk opened the pull request:
To allow users to reshare posts from the post list screen, we want to bring them to the editor but with the share post modal open.

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Change the `jetpackSidebarIsOpen` boolean query param to `jetpackSidebar` enum with values being `open` or `open_with_share_post` to avoid multiple query params.
* Create a re-usable utility function to handle the loagic for opening Jetpack sidebar with URL query param
* Open share modal on editor load via URL query param
* Remove the jetpack sidebar logic from Jetpack plugin as it's not used there anymore and we have everything in publicize now.
* Fix react warning about uncontrolled component being controlled

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to post list page
* Confirm that you see the "Share" link for published posts but not for drafts
* Click on the link
* Confirm that it goes to the editor with share post modal open
* Try the above with only Jetpack and Social activated individually and together.

manzoorwanijk committed:
Create a re-usable utility function for opening Jetpack sidebar

manzoorwanijk committed:
Use the shared function in Jetpack and Social sidebar

manzoorwanijk committed:
Update post list logic to use the updated query param

manzoorwanijk committed:
Enable share link only for published posts

manzoorwanijk committed:
Enable share link for all posts

manzoorwanijk committed:
Open share modal on editor load

manzoorwanijk committed:
Fix react warning about uncontrolled component being controlled

manzoorwanijk committed:
Remove the jetpack sidebar logic from Jetpack plugin

manzoorwanijk committed:
Add changelog

manzoorwanijk commented on projects/js-packages/publicize-components/src/components/social-post-modal/preview-section.tsx:
```diff
@@ -79,7 +79,7 @@ export function PreviewSection() {
 			<TabPanel tabs={ connections }>
 				{ ( tab: ( typeof connections )[ number ] ) => {
-					const isEnabled = canBeTurnedOn( tab ) && tab.enabled;
+					const isEnabled = Boolean( canBeTurnedOn( tab ) && tab.enabled );
```
This is to fix this warning

<img width="960" alt="Screenshot 2025-03-11 at 3 59 05 PM" src="https://github.com/user-attachments/assets/91eb0c1d-213b-4907-9ce1-47c7d84d35ca" />

manzoorwanijk committed:
Remove query param from URL on modal close

pablinos commented on projects/js-packages/publicize-components/src/components/block-editor/shared-utils.ts:
```diff
@@ -0,0 +1,27 @@
+import { store as interfaceStore } from '@wordpress/interface';
+import { getQueryArg } from '@wordpress/url';
+
+const JETPACK_SIDEBAR_OPTIONS = [ 'open', 'open_with_share_post' ];
```
I wonder if this is more than we need. `open` isn't going be used, and any other actions might not be sidebar related.

We could make the parameter something more generic like `jetpack-editor`, and have a value like `share_post`, or leave the parameter as `jetpack-sidebar` but just have the one value.

manzoorwanijk replied:
Updated in fc023652b82730952cd3f74912521c0596948a04

pablinos commented on projects/js-packages/publicize-components/src/components/block-editor/shared-utils.ts:
```diff
@@ -0,0 +1,27 @@
+ */
+export function mayBeOpenSidebar( sidebar: 'jetpack' | 'social' ) {
+	domReady( () => {
+		const value = getQueryArg( window.location.search, 'jetpackSidebar' ) as string;
```
Nitpick: I'm not sure why we were using camel case before. Perhaps we should take the opportunity to change that.

manzoorwanijk replied:
Updated in fc023652b82730952cd3f74912521c0596948a04

pablinos commented on projects/js-packages/publicize-components/src/components/social-post-modal/modal.tsx:
```diff
@@ -15,7 +16,19 @@ import styles from './styles.module.scss';
-	const [ isModalOpen, toggleModal ] = useReducer( state => ! state, false );
+	const [ isModalOpen, toggleModal ] = useReducer(
+		state => {
+			const url = new URL( window.location.href );
+			// If the modal is opened with the share post query arg, remove it from the URL.
+			if ( state && url.searchParams.has( 'jetpackSidebar' ) ) {
+				url.searchParams.delete( 'jetpackSidebar' );
+				window.history.replaceState( null, '', url.toString() );
+			}
+			return ! state;
```
Do we have to remove it? It's probably cleaner, but it would be better if this didn't live here. I also think it's fine to leave it in the URL so if the page is reloaded, it remains open

manzoorwanijk replied:
If we don't remove it, switching to some other settings sidebar and switching back to Jetpack will open the modal again, which can be annoying.

pablinos replied:
That points further to the state being in the wrong place. I was thinking that the initial state of the modal should be passed in and that we shouldn't be coupling the modal to this queryArg logic.

manzoorwanijk replied:
I have moved the logic to the store. Thanks

pablinos commented on projects/js-packages/publicize-components/src/components/social-post-modal/modal.tsx:
```diff
@@ -15,7 +16,19 @@ import styles from './styles.module.scss';
+			return ! state;
+		},
+		null,
+		() => getQueryArg( window.location.search, 'jetpackSidebar' ) === 'open_with_share_post'
```
I wonder if this should be here also. I think it might be better to do all the query parameter checking in one place, and perhaps we could pass in the initial state of the modal?

manzoorwanijk replied:
Done in 3dd0fe7326db1d4a0886ed2d69b017b4cb5a6cbe

Thanks

pablinos replied:
I'm not sure that's quite what I was thinking. I'm not sure that this component should need to use a utility that looks at the query args. The [JetpackSidebarManager](https://github.com/Automattic/jetpack/commit/3dd0fe7326db1d4a0886ed2d69b017b4cb5a6cbe#diff-d9ce49b8a840e0e689a75d525336bef1a87e4ddcb8fd9cf2d507a2dec7810bd3R8-R20) seems too specific also. We're not just thinking about the sidebar. I think we want something that triggers an action based on a query parameter when the editor loads, that way we can add other things in future, perhaps to trigger tours, highlight something in the UI, etc. And if the initial state is managed outside of the component, we'll not need the removeArg method.

manzoorwanijk replied:
Ah OK, that means we have to make that modal state global.

pablinos commented on projects/packages/publicize/src/class-publicize-setup.php:
```diff
@@ -97,6 +99,29 @@ public static function register_core_options() {
+	 * Things to do on admin_init.
+	 */
+	public static function admin_init_actions() {
+		( new Jetpack_Social_Settings\Dismissed_Notices() )->register();
```
This will already be run as part of register_core_options

manzoorwanijk replied:
Removed in e6e84d69d5fb97bd7dc2f9dd8dd89b29cc4b85f6.

Thanks

pablinos commented on projects/packages/publicize/src/class-publicize-setup.php:
```diff
@@ -97,6 +99,29 @@ public static function register_core_options() {
+	 */
+	public static function admin_init_actions() {
+		( new Jetpack_Social_Settings\Dismissed_Notices() )->register();
+		add_action( 'current_screen', array( self::class, 'add_filters_and_actions_for_screen' ), 5 );
```
`current_screen` is only run in the admin, so perhaps we could add this action directly in `on_jetpack_feature_publicize_enabled` rather than in an admin_init hook.

manzoorwanijk replied:
Done in e6e84d69d5fb97bd7dc2f9dd8dd89b29cc4b85f6.

Thanks

pablinos reviewed (commented):
I don't see the Share link until I enable the Post_List module with:

`add_action( 'admin_init', array( '\Automattic\Jetpack\Post_List\Post_List', 'configure' ) );`

Should we add the post action without Post_List just from Publicize?

But with this it works well. I had a few comments.

manzoorwanijk committed:
Merge branch 'trunk' into update/social/open-share-post-modal-via-url

manzoorwanijk commented:
> I don't see the Share link until I enable the Post_List module with:
> 
> `add_action( 'admin_init', array( '\Automattic\Jetpack\Post_List\Post_List', 'configure' ) );`
> 
> Should we add the post action without Post_List just from Publicize?
> 
> But with this it works well. I had a few comments.

I think we can do that as a separate change.

manzoorwanijk committed:
Rename the param to `jetpack-sidebar` with just one value - `share_post`

manzoorwanijk committed:
Directly hook into current_screen without admin_init

manzoorwanijk committed:
Do all the query parameter checking in one place

pablinos reviewed (commented):
What about adding the modal state to the store? We could define its initial state from the query parameter and then pass in `isModalOpen` as a prop.

manzoorwanijk committed:
Store the share post modal state in the store

manzoorwanijk committed:
Update JetpackSidebarManager manager to dispatch store action

manzoorwanijk committed:
Update modal to use the state from the store

manzoorwanijk commented:
> What about adding the modal state to the store? We could define its initial state from the query parameter and then pass in `isModalOpen` as a prop.

I have moved the state to the store and the modal component looks cleaner now. Thanks

pablinos commented on projects/js-packages/publicize-components/src/components/social-post-modal/modal.tsx:
```diff
@@ -1,57 +1,75 @@
+ * @return - Social Post Modal component.
+ */
+function RenderSocialPostModal( { title }: { title: string } ) {
+	const { closeSharePostModal } = useDispatch( socialStore );
```
If this is going to be another component, shall we pass this in? But equally we could leave this code where it is for now.

manzoorwanijk replied:
I have moved the logic to props.

pablinos commented on projects/js-packages/publicize-components/src/components/block-editor/shared-utils.ts:
```diff
@@ -0,0 +1,65 @@
+export class JetpackSidebarManager {
+	url: URL;
+
+	queryArg = 'jetpack-sidebar';
```
I'm not sure we need all of this now, and it's not really about the sidebar. What about adding a function to the shared utils package, which gets the value a more generic parameter like `jetpack-editor-action`, and then if that equals `reshare_post` it opens the modal?

manzoorwanijk replied:
Done.

pablinos commented on projects/js-packages/publicize-components/src/components/block-editor/shared-utils.ts:
```diff
@@ -0,0 +1,65 @@
+	 *
+	 * @param sidebar - The sidebar to open
+	 */
+	handleSidebarQuery( sidebar: 'jetpack' | 'social' ) {
+		domReady( () => {
+			const queryArg = this.getQueryArg();
+			if ( queryArg ) {
+				const area = sidebar === 'jetpack' ? this.jetpackSidebar : this.socialSidebar;
+
+				const { enableComplementaryArea } = dispatch( interfaceStore ) as {
+					enableComplementaryArea: ( scope: string, _area: string ) => void;
+				};
+
+				enableComplementaryArea( 'core', area );
+
+				if ( queryArg === 'share_post' ) {
+					dispatch( socialStore ).openSharePostModal();
+				}
+
+				// Remove the query arg from the URL
+				this.removeQueryArg();
+			}
+		} );
```
Based on my comment above this then becomes a single function, and we don't need to remove the argument from the URL

manzoorwanijk replied:
Done

manzoorwanijk committed:
Move the generic logic to shared utils package

manzoorwanijk committed:
Move the modal close logic to props

manzoorwanijk committed:
Fix dom-ready dependency

manzoorwanijk committed:
Fix pnpm build-extensions

pablinos commented on projects/js-packages/shared-extension-utils/src/block-editor-actions.js:
```diff
@@ -0,0 +1,55 @@
+ *
+ * @return {string | null} The Jetpack Editor action.
+ */
+export function getJetpackEditorAction() {
+	const url = new URL( window.location.href );
+
+	return url.searchParams.get( JETPACK_EDITOR_ACTION );
+}
```
I'm probably overthinking this and confusing the situation, but I was thinking that this function would be the only thing we'd add here.

pablinos commented on projects/js-packages/shared-extension-utils/src/block-editor-actions.js:
```diff
@@ -0,0 +1,55 @@
+ *
+ * @param {{sidebarToOpen?: string; onAction?: (action:string) => void; removeQueryArg?: boolean}} args - Arguments.
+ */
+export function handleJetpackEditorAction( {
```
...but I quite like this, so the domReady stuff is contained here.

pablinos commented on projects/js-packages/shared-extension-utils/src/block-editor-actions.js:
```diff
@@ -0,0 +1,55 @@
+ * @param {{sidebarToOpen?: string; onAction?: (action:string) => void; removeQueryArg?: boolean}} args - Arguments.
+ */
+export function handleJetpackEditorAction( {
+	sidebarToOpen = 'jetpack-sidebar/jetpack',
```
However, I don't think this should be concerned with sidebars.

manzoorwanijk replied:
I have moved that sidebar logic back to publicize-components.

pablinos commented on projects/js-packages/shared-extension-utils/src/block-editor-actions.js:
```diff
@@ -0,0 +1,55 @@
+} ) {
+	domReady( () => {
+		const action = getJetpackEditorAction();
+		if ( action ) {
```
I think we should pass in the action param we're looking for and a callback. If `action` equals the value we call the callback.

manzoorwanijk replied:
Done in f19e968463a2027f10d7a48c35b84c2495892b46

pablinos reviewed (commented):
I think I've sent you down a rabbit hole so we're over-engineering it now! I was trying to make sure that we only added what we need right now. I'm still not sure if we need to be removing the query parameter for example.

But if we're going to remove this once we have the modal in the post list screen, then perhaps we should move it all back into publicize-components, as it will only be short lived.

manzoorwanijk committed:
Change the logic to callback on action

manzoorwanijk committed:
Do not remove query arg by default

manzoorwanijk commented:
I have implemented your suggestions.

> I'm still not sure if we need to be removing the query parameter for example.

I have changed the logic to not remove it by default, unless the callback implies so.

pablinos commented:
> if the user closes the moda

That's probably the key. We could remove the argument then if you like

manzoorwanijk committed:
Remove query param on modal close

pablinos reviewed (approved):
Works well!
